### PR TITLE
Chapter 3 challenges

### PR DESCRIPTION
**Pull request by OnicaAdrianCatalin**
## New features

Answering a question will disable answer buttons and enable next button.

After answering all questions you will get the result.

## Changes in code

Added new parameter in Question data class to check if the question was answered.

Added validation, if you answered a question the answer buttons will be disabled and next button will be enabled. This will prevent user from not answering all the questions.

Added validation, for answering all questions the percentage score will be displayed in a toast.

https://user-images.githubusercontent.com/102966412/162924607-81afa24f-84f4-48f2-8b2a-3d75043613dc.mov

### CODE REVIEW

File: app/src/main/java/com/example/geoquiz/MainActivity.kt [hunk 1]
```diff
@@ -39,10 +43,20 @@ class MainActivity : AppCompatActivity() {
     private fun setOnClickListeners() {
         trueButton.setOnClickListener {
             checkAnswer(true)
+            questionBank[currentIndex].answeredQuestion = true
+            answeredQuestionCheck()
+            if (questionBank[questionBank.size - 1].answeredQuestion == true) {
```
Comment (ghost) [inline app/src/main/java/com/example/geoquiz/MainActivity.kt]: `if` statement can be simplified with `questionBank[questionBank.size - 1].answeredQuestion`

File: app/src/main/java/com/example/geoquiz/MainActivity.kt [hunk 2]
```diff
@@ -39,10 +43,20 @@ class MainActivity : AppCompatActivity() {
...
+                getCorrectAnswerPercentage()
+            }
         }
 
         falseButton.setOnClickListener {
             checkAnswer(false)
+            questionBank[currentIndex].answeredQuestion = true
+            answeredQuestionCheck()
+            if (questionBank[questionBank.size - 1].answeredQuestion == true) {
```
Comment (ghost) [inline app/src/main/java/com/example/geoquiz/MainActivity.kt]: same here

File: app/src/main/java/com/example/geoquiz/MainActivity.kt [hunk 3]
```diff
@@ -51,46 +65,52 @@ class MainActivity : AppCompatActivity() {
...
-    override fun onPause() {
-        super.onPause()
-        Log.d(tag, "onPause called ")
-    }
-
-    override fun onStop() {
-        super.onStop()
-        Log.d(tag, "onStop called ")
+    private fun answeredQuestionCheck() {
+        var checkAnswer = questionBank[currentIndex].answeredQuestion
```
Comment (ghost) [inline app/src/main/java/com/example/geoquiz/MainActivity.kt]: `checkAnswer` can be a `val`

File: app/src/main/java/com/example/geoquiz/MainActivity.kt [hunk 4]
```diff
@@ -51,46 +65,52 @@ class MainActivity : AppCompatActivity() {
...
+        if (checkAnswer == false) {
```
Comment (ghost) [inline app/src/main/java/com/example/geoquiz/MainActivity.kt]: can be simplified to `if (!checkAnswer)`

File: app/src/main/java/com/example/geoquiz/MainActivity.kt [hunk 5]
```diff
@@ -51,46 +65,52 @@ class MainActivity : AppCompatActivity() {
...
+            trueButton.setEnabled(true)
```
Comment (ghost) [inline app/src/main/java/com/example/geoquiz/MainActivity.kt]: You can use property access syntax here: `trueButton.isEnabled = true`

File: app/src/main/java/com/example/geoquiz/MainActivity.kt [hunk 6]
```diff
@@ -51,46 +65,52 @@ class MainActivity : AppCompatActivity() {
...
+            trueButton.setEnabled(false)
+            falseButton.setEnabled(false)
+        }
     }
 
-    override fun onDestroy() {
-        super.onDestroy()
-        Log.d(tag, "onDestroy called ")
+    private fun getCorrectAnswerPercentage() {
+        var answersPercentage = 0
```
Comment (ghost) [inline app/src/main/java/com/example/geoquiz/MainActivity.kt]: This can be removed and declared directly inside the `if` statement with `val`

**Review by ghost — COMMENTED** [review]
Looks good, just needs some minor clean up here and there.
I have a question about the previous button, when the user taps it should he be able to answer the previous question again?

**Comment by OnicaAdrianCatalin** [comment]
When user presses the previous button he should not be able to answer again the question, answer buttons should be disabled.

**Review by ghost — APPROVED** [review]

File: app/src/main/java/com/example/geoquiz/MainActivity.kt [hunk 7]
```diff
@@ -54,46 +65,52 @@ class MainActivity : AppCompatActivity() {
...
-
-    override fun onPause() {
-        super.onPause()
-        Log.d(TAG, "onPause called ")
-    }
-
-    override fun onStop() {
-        super.onStop()
-        Log.d(TAG, "onStop called ")
+    private fun answeredQuestionCheck() {
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/geoquiz/MainActivity.kt]: Based on the name I assume the function to check all answered questions.
Instead, it sets the button states based on the current question's answered state.
I would rename to `updateQuestionButtonsState(isCurrentQuestionAnswered: Boolean)` - and pass `questionBank[currentIndex].answeredQuestion` as param

File: app/src/main/java/com/example/geoquiz/MainActivity.kt [hunk 8]
```diff
@@ -54,46 +65,52 @@ class MainActivity : AppCompatActivity() {
...
+            nextButton.isEnabled = true
+            trueButton.isEnabled = false
+            falseButton.isEnabled = false
+        }
     }
 
-    override fun onDestroy() {
-        super.onDestroy()
-        Log.d(TAG, "onDestroy called ")
+    private fun getCorrectAnswerPercentage() {
```
Comment (NegreaVlad) [inline app/src/main/java/com/example/geoquiz/MainActivity.kt]: Good name, but all functions should have a single responsibility and all methods named get should return something.
This one computes the percentage and shows the toast. So it is not single responsibility.

Remove the toast from here, make it return the percentage and show the toast in the click listener

File: app/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -1,11 +1,13 @@
 <resources>
-    <string name="app_name">QuizApplication</string>
+    <string name="app_name">QuizApplication
```
Comment (NegreaVlad) [inline app/src/main/res/values/strings.xml]: Should be on a single line

**Review by NegreaVlad — APPROVED** [review]
Looks good!

**Pull request merged**